From c9132973f88015586aa847ffcaa96e86bb23776f Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Fri, 4 Jul 2025 08:14:55 +0100
Subject: [PATCH] file: trim: Don't try BLKDISCARD if earlier
 FALLOC_FL_PUNCH_HOLE worked

In file_zero, we test if the operations we are trying succeed and if
so jump to a single 'out:' label where we deal with the success path
out of that function.

We did not do the same thing in file_trim.  Thus in the case where
FALLOC_FL_PUNCH_HOLE succeeds, we might fall through to trying
BLKDISCARD as well.  As it happens we probably don't do this (at
least, in Linux) because we only try BLKDISCARD for block devices, and
FALLOC_FL_PUNCH_HOLE does not work on those.  But it's a good thing to
clean up this code anyway, especially if we were to add more cases in
future.

This also adds a debug message if none of the trim methods worked,
which is also analogous to what happens in the same part of file_zero.

(cherry picked from commit 909e483c121c69e6b2759ef9d5401eb3d5acc998)
---
 plugins/file/file.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/plugins/file/file.c b/plugins/file/file.c
index b881da37..66d03d4f 100644
--- a/plugins/file/file.c
+++ b/plugins/file/file.c
@@ -1191,6 +1191,7 @@ file_trim (void *handle, uint32_t count, uint64_t offset, uint32_t flags)
 
     r = do_fallocate (h->fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE,
                       offset, count);
+    if (r == 0) goto out;
     if (r == -1) {
       if (!is_enotsup (errno)) {
         nbdkit_error ("fallocate: %s: offset=%" PRIu64 ", count=%" PRIu32 ":"
@@ -1214,6 +1215,7 @@ file_trim (void *handle, uint32_t count, uint64_t offset, uint32_t flags)
     uint64_t range[2] = {offset, count};
 
     r = ioctl (h->fd, BLKDISCARD, &range);
+    if (r == 0) goto out;
     if (r == -1) {
       if (!is_enotsup (errno)) {
         nbdkit_error ("ioctl: %s: offset=%" PRIu64 ", count=%" PRIu32 ":"
@@ -1227,6 +1229,15 @@ file_trim (void *handle, uint32_t count, uint64_t offset, uint32_t flags)
   }
 #endif
 
+  /* Trim is advisory.  If we got here, we were unable to trim. */
+  ONCE (nbdkit_debug ("%s: could not trim, no trim methods worked",
+                      h->name));
+  return 0;
+
+#ifdef __clang__
+  __attribute__ ((unused))
+#endif
+ out:
   if ((flags & NBDKIT_FLAG_FUA) && file_flush (handle, 0) == -1)
     return -1;
 
-- 
2.47.1

